examples/vision/semisupervised_simclr.py

```diff
@@ -0,0 +1,743 @@
+
+"""
+## Dataset
+
+During training we will load a large batch of unlabeled images along with a smaller batch
+of labeled images simultaneously.
+"""
+
+
+def prepare_dataset(steps_per_epoch):
+    # labeled and unlabeled samples are loaded synchronously
+    # with batch sizes selected accordingly
+    unlabeled_batch_size = 100000 // steps_per_epoch
+    labeled_batch_size = 5000 // steps_per_epoch
+    batch_size = unlabeled_batch_size + labeled_batch_size
+    print(
+        "batch size is {} (unlabeled) + {} (labeled)".format(
+            unlabeled_batch_size, labeled_batch_size
+        )
+    )
+
+    unlabeled_train_dataset = (
+        tfds.load("stl10", split="unlabelled", as_supervised=True, shuffle_files=True)
+        .shuffle(buffer_size=5000)
+        .batch(unlabeled_batch_size, drop_remainder=True)
+    )
+    labeled_train_dataset = (
+        tfds.load("stl10", split="train", as_supervised=True, shuffle_files=True)
+        .shuffle(buffer_size=5000)
+        .batch(labeled_batch_size, drop_remainder=True)
+    )
+    test_dataset = (
+        tfds.load("stl10", split="test", as_supervised=True)
+        .batch(batch_size)
+        .prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
+    )
+
+    # labeled and unlabeled datasets are zipped together
+    train_dataset = tf.data.Dataset.zip(
+        (unlabeled_train_dataset, labeled_train_dataset)
+    ).prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
+
+    return batch_size, train_dataset, labeled_train_dataset, test_dataset
+
+
+# load STL10 dataset
+batch_size, train_dataset, labeled_train_dataset, test_dataset = prepare_dataset(
+    steps_per_epoch
+)
+
+"""
+## Custom image augmentation layers
+
+The two most important image augmentations for contrastive learning are the following:
+- cropping: forces the model to encode different parts of the same image similarly
+- color jitter: prevents a trivial color histogram-based solution to the task by
+distorting color histograms
+
+In this example we use random horizontal flips as well.
+
+We implement our image augmentations as custom preprocessing layers. This has the
+following two advantages:
+- the data augmentation will run on the GPU, so the training will not be bottlenecked by
+the data pipeline in environments with constrained CPU resources (such as a Colab
+Notebook, or a personal machine)
+- deployment is easier as the data preprocessing pipeline is encapsulated in the model,
+and does not have to be reimplemented when deploying it
+"""
+
+# the implementation of these image augmentations follow the torchvision library:
+# https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py
+# https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional_tensor.py
+
+# however these augmentations:
+# -run on batches of images
+# -run on gpu
+# -can be part of a model
+
+
+# crops and resizes part of the image to the original resolutions
+class RandomResizedCrop(layers.Layer):
```

Comment:
You can remove this layer and instead use a combination of the layers RandomCrop and Resizing. Saves you a good number of lines